### PR TITLE
Introduce a ResultAsync type for better async Results

### PR DESCRIPTION
**Pull request by paduc**
This is the PR for issue #32 

I'm waiting for your confirmation of the new API before updating the docs.

### CODE REVIEW

File: src/result-async.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
+import { Result, Ok, Err } from './result'
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isPromise<T>(obj: any): obj is Promise<T> {
```
Comment (supermacro) [inline src/result-async.ts:4]: Please use `unknown` instead of `any` for function argument. 

I think it would be beneficial to have a few tests for this function. Are there any edge cases that you think that should be tested?

I was initially thinking it would be beneficial to use a 3rd party library, but that would add unjustified bloat to this package. But now that there's a custom-made `isPromise` validator, it would be prudent to ensure it's working as expected (plus to verify regressions in the future).

Comment (paduc) [reply src/result-async.ts:4]: Hey, this method was copied from this repo: [then/is-promise](https://github.com/then/is-promise).

The reason I didn't import it is because of the risks associated with dependencies. Very recently, the `is-promise` repo (which is listed as a dependency to 3.4M projects!) introduced a breaking change which wreaked havoc across a huge number of dependants. 

I think it's trivial enough to just paste here. This implementation coincides with the specification of a `thenable`: an `object` or `function` with a `then` property which is a `function`. It's a looser alternative to "instanceof Promise" which returns false for promises coming from other libs like Bluebird for instance. I don't see any edge-cases and if 3.4M projets use it, I guess there aren't any.

I'm not sure if having tests for this method is necessary to be honest. If you prefer, we can copy those located [here](https://github.com/then/is-promise/blob/master/test.js).

Comment (paduc) [reply src/result-async.ts:4]: As for the `unknown` argument type, I tried but I have a compiler error on the `obj.then` section: `Property 'then' does not exist on type 'object'`.

**Comment by paduc** [comment]
> Hey @paduc unfortunately I ran out of time this morning when reading through your PR, so I haven't really looked at much. Basically the only thing I've looked at thoroughly is `isPromise`. Please give me a few days in order to properly do code review.

No problem, take your time ! There aren't many lines of code but the implications aren't obvious... 

The principle is this: each `ResultAsync` instance holds a `Promise<Result>`.
For each chaining event (`map/mapErr/andThen`), the inner promise is chained with a `.then` which handles the different cases (Error/Value) and returns the appropriate `Result`, resulting in a new `Promise<Result>` that is then wrapped inside a new `ResultAsync` instance which is returned.

The `then` class method makes `ResultAsync` awaitable. It transforms a `ResultAsync` back into a `Promise<Result>`. 

One additional aspect: each chaining method accepts asynchronous function arguments. So the map function might return either a value (which is then wrapped in `Ok`) or a promise (which has to be unpacked with a `.then`).
A beauty of promises is that they are flattened automatically: returning an instance of `Result` or `Promise<Result>` inside a `then` callback will always result in a `Promise<Result>`.

Don't hesitate if you have further questions/remarks !

File: src/result-async.ts (hunk 1 above)
Comment (supermacro) [reply src/result-async.ts:4]: Yup, I agree it makes sense not to use a third-party lib. But for the sake of tracking regressions, there should definitely be some tests for this function.

I tried getting rid of `any`, and got as far as [this](https://www.typescriptlang.org/play/#code/C4TwDgpgBACgNgQwJYDsDyAjAVhAxsKAXigCU8B7AJwBMAeAZ2EtQHMAaKAVxQGsVyA7igB8AKFEAzbviTkUUJPXjIUAKXqYc+ABQA3BHABcXXvyEBKY-rgL6sRKk14CAb1FQolCME6V5oSHIJKGsiQmIAcnJsZwioADJ4kIMoAEJwqBROOBtRAF8ocUlpYFl5RRhKcgBbRQhaABVhbWisY24+QRRLKFbbWCra+nqmqDcPLx8-KG13DzTUvsS5j20KhzUNGJ1W8ygAH32oAIgg3uwwyKkUGTkIveX54-BT4NaAOmAACwh5cKuSmUInNzPlREA) ... but I am still getting type errors and so I think it's justified to keep the `any` in this case.

Comment (paduc) [reply src/result-async.ts]: I agree about the tests. I just pushed a commit with extensive tests.

File: src/is-promise.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const isPromise = <T>(obj: any): obj is Promise<T> => {
```
Comment (supermacro) [inline src/is-promise.ts:2]: For the sake of consistency with the rest of the package, please`export` here instead of exporting an object at the end of the file.

File: src/result-async.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+export class ResultAsync<T, E> {
+  private _promise: Promise<Result<T, E>>
+
+  constructor(res: Promise<Result<T, E>>) {
+    this._promise = res
+  }
+
+  static fromPromise<T, E>(promise: Promise<T>): ResultAsync<T, E> {
+    return new ResultAsync(promise.then((value: T) => new Ok<T, E>(value)))
```
Comment (supermacro) [inline src/result-async.ts:12]: What happens if calling `promise.then` results in a rejected promise? It would be odd if neverthrow started throwing unhandled exceptions 🙈 !
  - What if there was a `catch` here that returned an `Err`?

File: src/result-async.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      }
+
+      const newValue = _f(res.value)
+
+      if (isPromise(newValue)) {
+        return newValue.then(value => new Ok<A, E>(value))
+      }
+
+      return new Ok<A, E>(newValue)
+    })
```
Comment (supermacro) [inline src/result-async.ts:28]: Looks great! I have a few thoughts. 

For the sake of keeping code as legible as possible:
- Any reason not to use `async` / `await` syntax? It might lead to less indentation
- Please omit explicit type annotations if the compiler can infer the type already

The code on line 16 would look like this:  

```typescript
    const newPromise = this._promise.then((result) => { ...
```

Secondly, the instantiation of `Err` and `Ok` values on line 18, 24 and 27 shouldn't need to be annotated with their respective type arguments either. See if you can omit them. The compiler should be able to infer the types in all of these cases. 

If you do go ahead with these changes, please apply them to all the methods below as well! Otherwise please respond with some counterarguments so we can agree on a consistent coding style!

File: src/result-async.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    return new ResultAsync(newPromise)
+  }
+
+  match = <A>(ok: (t: T) => A, _err: (e: E) => A): Promise<A> => {
+    return this._promise.then(res => res.match(ok, _err))
+  }
+
+  // Makes ResultAsync awaitable
+  then(_successCallback: (res: Result<T, E>) => void) {
```
Comment (supermacro) [inline src/result-async.ts:74]: Cool! I learned something new :) By the way, does this method need to return anything? Seems counter-intuitive that this method doesn't return a value. 

In my head I am thinking, "in normal promises, I can chain `.then` as many times as I want ... so wouldn't that imply that a `.then` method returns a `Promise`?"

File: src/result-async.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    this._promise.then(_successCallback)
+  }
+}
+
+export const okAsync = <T, E>(value: T): ResultAsync<T, E> =>
+  new ResultAsync(Promise.resolve(new Ok<T, E>(value)))
+
+export const errAsync = <T, E>(err: E): ResultAsync<T, E> =>
+  new ResultAsync(Promise.resolve(new Err<T, E>(err)))
```
Comment (supermacro) [inline src/result-async.ts:83]: What's the use case for having `okAsync` and `errAsync`? It seems to me that if I want to convert a normal result into a async result, I would call `.asyncMap`, no?

I would prefer to keep the surface area of code as small as possible. This is where I'm coming from when I ask this question, and If there's no good reason right now to keep `okAsync` and `errAsync`, then I would ask that it be removed.

File: src/result-async.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+  constructor(res: Promise<Result<T, E>>) {
+    this._promise = res
+  }
+
+  static fromPromise<T, E>(promise: Promise<T>): ResultAsync<T, E> {
+    return new ResultAsync(promise.then((value: T) => new Ok<T, E>(value)))
+  }
+
+  map<A>(_f: (t: T) => A | Promise<A>): ResultAsync<A, E> {
```
Comment (supermacro) [inline src/result-async.ts:15]: change `_f` to `f` since it is being used in the body of this method. The convention that I've been abiding by is that if an argument is not being used then I signal it to the reader (and the compiler) by prepending the underscore.

Please follow this convention for all code in this PR.

Example:

https://github.com/supermacro/neverthrow/blob/master/src/result.ts#L77

**Comment by supermacro** [comment]
Hey @paduc ... not sure why the TravisCI status check didn't run, but I have now resolved that issue. 

You can ignore the Codacy status check (it's something I'm still evaluating and trying out). But the TravisCI check is required to pass before merging!

File: src/result-async.ts (hunk 3 above)
Comment (paduc) [reply src/result-async.ts]: You are totally right and I actually had already modified this function in another branch.
What I propose is to offer a second argument to `fromPromise` to handle the error case. This second argument should be a function which receives whatever is thrown by the promise and returns an `E` type value.

It looks like this now:
```typescript
  static fromPromise<T, E>(promise: Promise<T>, errorFn?: (e: unknown) => E): ResultAsync<T, E> {
    let newPromise: Promise<Result<T, E>> = promise.then((value: T) => new Ok(value))
    if (errorFn) {
      newPromise = newPromise.catch(e => new Err<T, E>(errorFn(e)))
    }
    return new ResultAsync(newPromise)
  }
```

File: src/result-async.ts (hunk 4 above)
Comment (paduc) [reply src/result-async.ts]: > * Any reason not to use `async` / `await` syntax? It might lead to less indentation

These methods need to be remain synchronous (as they are returning a `ResultAsync` instance) and `async` / `await` is only available inside asynchronous functions.

However, it's possible to pass an asynchronous function to `.then()` and `await` the value instead of chaining with an addition `.then()`. 
I also noticed, we can `await` non-promise values as well. Given we don't know if the provided function returns a promise or not, we can `await` in both cases.

Looks a lot cleaner like this:
```typescript
  map<A>(f: (t: T) => A | Promise<A>): ResultAsync<A, E> {
    const newPromise = this._promise.then(async (res: Result<T, E>) => {
      if (res.isErr()) {
        return new Err<A, E>(res.error)
      }

      return new Ok<A, E>(await f(res.value))
    })

    return new ResultAsync(newPromise)
  }
```

Added bonus: we don't need `isPromise` at all anymore 🎉

File: src/result-async.ts (hunk 7 above)
Comment (paduc) [reply src/result-async.ts]: Of course, I misinterpreted your conventions. I fixed this.

File: src/result-async.ts (hunk 6 above)
Comment (paduc) [reply src/result-async.ts:70]: Here's an example use-case:

```typescript
const createUser = (login: string): ResultAsync<boolean, Error> => {
  if (login.length < 3) {
    return errAsync(new Error('Username is too short'))
  }

  return ResultAsync.fromPromise(
    insertUserIntoDb(login),
    () => new Error('Failed to insert into database'),
  )
}
```

=> A method that should return a `ResultAsync` and also performs synchronous operations. It's the async equivalent of `Result`s `ok` and `err` handlers.
It avoids writing `return new ResultAsync(Promise.resolve(new Err<T, E>(err)))`...

File: src/result-async.ts (hunk 5 above)
Comment (paduc) [reply src/result-async.ts]: You are absolutely correct.

I added a test (which of course didn't pass with the previous version) and fixed the implementation.

File: src/result-async.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    this._promise = res
+  }
+
+  static fromPromise<T, E>(promise: Promise<T>, errorFn?: (e: unknown) => E): ResultAsync<T, E> {
+    let newPromise: Promise<Result<T, E>> = promise.then((value: T) => new Ok(value))
+    if (errorFn) {
+      newPromise = newPromise.catch(e => new Err<T, E>(errorFn(e)))
+    }
+    return new ResultAsync(newPromise)
+  }
```
[the remark this answers is not in the document]
Comment (paduc) [reply src/result-async.ts:16]: Good idea !

It's a bit tricky though because `process.env` is only for `node` runtimes. Isn't this lib also for the browser ?

We could either add a condition to check if it's a browser environment (would we want to log in the browser as well ?).

We could also make the second argument mandatory.

[the remark this answers is not in the document]
Comment (supermacro) [reply src/result-async.ts]: @paduc good point. I was thinking that `process.env` would be available because of webpack or parcel.

I think this warning should be logged anywhere. The reason I didn't want to log in production is because the user might already be dealing with promise rejections on their own. So it would be annoying to keep logging the warning even though it's irrelevant to them.

I don't think making it mandatory makes sense because users may prefer to catch the exception and return an `Err` themselves.

[the remark this answers is not in the document]
Comment (paduc) [reply src/result-async.ts]: Ok to recap, the solution would be:
- In node, only log outside of prod and test environments
- In the browser, always log ?

What if we only log if the second argument is `undefined` (meaning, no warning if the user provides `null` for instance ?

[the remark this answers is not in the document]
Comment (supermacro) [reply src/result-async.ts]: Yes, let's do that. Log outside of tests or prod for node. Always log in a browser environment (although some front end apps might have `process.env` defined, like webpack projects).

Don't add the `undefined` vs `null` functionality. I don't think it's worth it.

**Comment by paduc** [comment]
Hello @supermacro,

I'm working on the documentation and I am wondering about the `chain` methodology.
Obviously, you'll want to keep it to avoid breaking changes and also because it's a nice alternative to chaining.

However, you might want to address the fact that there are two ways of handling asynchronous behaviour now.

File: README.md [hunk 9]
```diff
@@ -385,15 +425,267 @@ import { parseHeaders } 'imaginary-http-parser'
...
+myResult.isErr() // true
+```
 
 ---
+
+### `ResultAsync.fromPromise` (method)
+
+Transforms a `Promise<T>` into a `ResultAsync<T, E>`.
+
+The second argument handles the rejection case of the promise. If it is ommited, the code might throw.
```
Comment (supermacro) [inline README.md:491]:
```suggestion
The second argument handles the rejection case of the promise. If it is ommited, **the code might throw** because `neverthrow` does not know if the promise you are passing to `fromPromise` has any promise rejection logic associated to it (via a `.catch` method call or `catch (err) {}` block).
```

**Comment by supermacro** [comment]
Hey @paduc almost there!

I think what I'll do is remove the `chain` api & documentation in a breaking-change release when I get some time.

I don't know when (or if) I will do that, but that's what I am thinking. So right now it's best to leave the `chain` docs as they are.

I noticed that travis / github are bugging out, so I'll simply run the CI checks on my machine after you've pushed those last documentation changes I requested, and if they pass I can merge and then publish a new version to npm 😁

Thanks for your patience! If you have any feedback on the whole process, feel free to share.

**Comment by paduc** [comment]
> Hey @paduc almost there!

Awesome ! Can't wait !

> 
> I think what I'll do is remove the `chain` api & documentation in a breaking-change release when I get some time.
> 
> I don't know when (or if) I will do that, but that's what I am thinking. So right now it's best to leave the `chain` docs as they are.
> 

Maybe we can add a line advising the use of `ResultAsync` and that it might be deprecated in the future ?

> I noticed that travis / github are bugging out, so I'll simply run the CI checks on my machine after you've pushed those last documentation changes I requested, and if they pass I can merge and then publish a new version to npm 😁
> 
> Thanks for your patience! If you have any feedback on the whole process, feel free to share.

This is my first meaningful open source contribution so I'd say the process is exciting and taught me a lot. I feel you've done a great job communicating your requirements and providing pointers !

File: README.md [hunk 10]
```diff
@@ -58,19 +54,61 @@ if (mappedYes.isOk()) {
...
 
 `Ok<T, E>`: contains the success value of type `T`
 
 `Err<T, E>`: contains the failure value of type `E`
 
+---
+
+### Asynchronous API
+
+Asynchronous methods can return a `ResultAsync` type instead of a `Promise<Result>` in order to enable further chaining.
```
[the remark this answers is not in the document]
Comment (paduc) [reply README.md:73]: I added a line to this effect. I also added an example using `.then()` below.

File: README.md [hunk 11]
```diff
@@ -7,18 +6,17 @@
 
 Encode failure into your program.
 
-This package contains a `Result` type that represents either success (`Ok`) or failure (`Err`).
-
-`neverthrow` also exposes an api for chaining sequential asynchronous tasks ([docs below](#chaining-api)).
+This package contains a `Result` type that represents either success (`Ok`) or failure (`Err`).  
+It also offers a `ResultAsync` type which wraps a `Promise<Result>` to enable chaining asynchronous tasks.
```
[the remark this answers is not in the document]
Comment (paduc) [reply README.md]: I added a line to this effect. What do you think ?

**Comment by supermacro** [comment]
Hey @paduc,

There's good news and bad news haha .. I approved the PR, but when I went to run the CI checks (which are `npm run lint`, `npm run typecheck` and `npm run test`), they unfortunately failed because of a missing `node` types package. See screenshot below. 

Screenshot from 2020-05-20 08-28-55

Please ensure that the above three npm scripts pass successfully on your machine.

Hopefully by running `npm install --save-dev @types/node` all your checks will pass and we can merge today.

**Comment by paduc** [comment]
> Hey @paduc,
> 
> There's good news and bad news haha .. I approved the PR, but when I went to run the CI checks (which are `npm run lint`, `npm run typecheck` and `npm run test`), they unfortunately failed because of a missing `node` types package. See screenshot below.
> 

My bad ! I had installed the `@types/node` locally but didn't commit the changes to `package.json` because I wasn't sure if we would keep the change yet.

I pushed a fix and all three commands run on my side.

**Comment by supermacro** [comment]
👍

**Comment by paduc** [comment]
🍾🥂
